pkg/apis/configuration/v1alpha1/kong_custom_entity_types.go

```diff
@@ -0,0 +1,94 @@
+package v1alpha1
+
+import (
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	KongCustomEntityKind = "KongCustomEntity"
+)
+
+type KongEntityScope string
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Namespaced,shortName=kce,categories=kong-ingress-controller,path=kongcustomentities,singular=kongcustomentity
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Entity Type",type=string,JSONPath=`.type`,description="type of the Kong entity"
+// +kubebuilder:validation:XValidation:rule="self.scope == oldSelf.scope",message="The scope field is immutable"
+// +kubebuilder:validation:XValidation:rule="self.type == oldSelf.type",message="The type field is immutable"
+// +kubebuilder:validation:XValidation:rule="!(self.type in ['services','routes','upstreams','targets','plugins','consumers','consumer_groups'])",message="The type field cannot be known Kong entity types"
+// +kubebuilder:validation:XValidation:rule="!(self.scope == 'independent' && has(self.parentRef)) && !(self.scope == 'attached' && !has(self.parentRef))",message="attached KongCustomEntity must have parentRef; independent KongCustomEntity must not have parentRef"
+// REVIEW: put all fields other than "status" under the "spec"?
+
+// KongCustomEntity defines a "custom" Kong entity that KIC cannot support the entity type directly.
+type KongCustomEntity struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Enum=independent;attached
+	// REVIEW: We can get the schema of entities from Kong gateway admin APIs including fields requiring foreign reference.
+	// Is the "Scope" field really required?
+
+	// Scope means whether the entity can be specified independently.
+	// "independent" means that the entity does not depend on other entities;
+	// "attached" means that the entity has "foreign" fields referring to other entities.
+	Scope KongEntityScope `json:"scope"`
+
+	// EntityType is the type of the Kong entity. The type is used in generating declarative configuration.
+	EntityType string `json:"type"`
+	// Fields defines the fields of the Kong entity itself.
+	Fields apiextensionsv1.JSON `json:"fields"`
+	// ControllerName specifies the controller that should reconcile it, like ingress class.
+	ControllerName string `json:"controllerName"`
+
+	// ParentRef references the kubernetes resource it attached to when its scope is "attached".
+	// Currently only KongPlugin/KongClusterPlugin allowed. This will make the custom entity to be attached
+	// to the entity(service/route/consumer) where the plugin is attached.
+	ParentRef *ObjectReference `json:"parentRef,omitempty"`
```

Comment:
For the "attached" entities (entities that have at lease one `foreign` field referring to other entities), we need a method to let KIC to know which other entit[y|ies] should it be attached to. In the design doc phase, we discussed how to attach custom entities to translated Kong entities (services/routes/consumers). We had several options:
- Like in this PR, define a `parentRef` in the `KongCustomEntity` CRD. Since the custom entities are brought by plugins, it is reasonable to let the `KongPlugin` to tell what entities should the entity to be attached to. 
- Define an annotation `konghq.com/custom-entities` in the resources that has the custom entities attached. For example, if we want a service to have a `degraphql_route` attached, we add the annotation `konghq.com/custom-entities` to the `Service` resource, then KIC will attach a `degraphql_route` to each of the Kong service that is translated from the referred `KongCustomEntity` in the annotation.
After some discussion, I decided to adopt option 1. What other options could be used to tell KIC what Kong entities should the custom entity to be attached to? @rainest

Reply:
Custom entities aren't inherently restricted to a specific instance of a plugin at the Kong level. If you have, for instance, a `key-auth` instance configured on IngressA and another on IngressB, a consumer with a `key-auth` credential can access either IngressA or IngressB (hence the usual need for the `acl` plugin alongside it).

As proposed we'd be auto-filling the Kong-side `service.id` in the custom entity with the generated service for whatever Service the KongPlugin is attached to? That seems a reasonable enough UX for custom entities like degraphql if we accept that it won't cover all custom entities, though it may get weird when a single Ingress generates multiple routes, each of which then (presumably) needs its own separate instance of the Kong-side entity.

You can probably just say "if you want to use this, restrict yourself to single-rule Ingresses/HTTPRoutes" to simplify things.